### PR TITLE
JIT: properly handle special-case varargs params in minopts

### LINKED ISSUE #19439: Assertion failed 'varDsc->lvRefCnt() == 0 && !varDsc->lvRegister && !varDsc->lvOnFrame'
From x86 Checked Windows R2R with JITMinOpts=1

https://ci.dot.net/job/dotnet_coreclr/job/master/view/x86/job/jitstress/job/x86_checked_windows_nt_r2r_jitminopts/135/consoleFull

```
JIT_Regression._CLR_x86_JIT_V1_M09_5_PDC_b28927_b28927_b28927_._CLR_x86_JIT_V1_M09_5_PDC_b28927_b28927_b28927_cmd [FAIL]
23:48:23         
23:48:23 
        Assert failure(PID 6908 [0x00001afc], Thread: 6328 [0x18b8]): Assertion failed 'varDsc->lvRefCnt() == 0 && !varDsc->lvRegister && !varDsc->lvOnFrame' in '_ldtoken:test(int):int:this' (IL size 4)
23:48:23         
23:48:23             File: d:\j\workspace\x86_checked_w---96237527\src\jit\regalloc.cpp Line: 427
23:48:23             Image: D:\j\workspace\x86_checked_w---96237527\bin\tests\Windows_NT.x86.Checked\Tests\Core_Root\crossgen.exe
23:48:23         
23:48:23         
23:48:23   
23:48:23   Return code:      1
23:48:23   Raw output file:      D:\j\workspace\x86_checked_w---96237527\bin\tests\Windows_NT.x86.Checked\Reports\JIT.Regression\CLR-x86-JIT\V1-M09.5-PDC\b28927\b28927\b28927.output.txt
23:48:23   Raw output:
23:48:23   BEGIN EXECUTION
23:48:23                 1 file(s) copied.
23:48:23         " D:\j\workspace\x86_checked_w---96237527\tests\..\bin\tests\Windows_NT.x86.Checked\Tests\Core_Root\crossgen.exe" /Platform_Assemblies_Paths D:\j\workspace\x86_checked_w---96237527\tests\..\bin\tests\Windows_NT.x86.Checked\Tests\Core_Root;D:\j\workspace\x86_checked_w---96237527\bin\tests\Windows_NT.x86.Checked\JIT\Regression\CLR-x86-JIT\V1-M09.5-PDC\b28927\b28927\IL;D:\j\workspace\x86_checked_w---96237527\bin\tests\Windows_NT.x86.Checked\JIT\Regression\CLR-x86-JIT\V1-M09.5-PDC\b28927\b28927 /in b28927.org /out b28927.exe
23:48:23         Microsoft (R) CoreCLR Native Image Generator - Version 4.5.30319.0
23:48:23         Copyright (c) Microsoft Corporation.  All rights reserved.
23:48:23         
23:48:23         Crossgen failed with exitcode - 123456789
23:48:23         Test Harness Exitcode is : 1
23:48:23         
23:48:23   To run the test:
23:48:23   > set CORE_ROOT=D:\j\workspace\x86_checked_w---96237527\bin\tests\Windows_NT.x86.Checked\Tests\Core_Root
23:48:23   > D:\j\workspace\x86_checked_w---96237527\bin\tests\Windows_NT.x86.Checked\JIT\Regression\CLR-x86-JIT\V1-M09.5-PDC\b28927\b28927\b28927.cmd
23:48:23   
```

Failing:
```
JIT\Regression\CLR-x86-JIT\V1-M09.5-PDC\b28927\b28927\b28927.cmd
JIT\Regression\CLR-x86-JIT\V1-M12-Beta2\b71120\b71120\b71120.cmd
```

### PR DESCRIPTION
**Pull request by AndyAyersMS**
We were previously marking all lcl vars as implicitly referenced in minopts,
but there are some special varargs params that can't ever be marked as
referenced, either implicitly or explicitly.

Special case these when computing or checking ref counts in minopts.

Closes #19439.

### CODE REVIEW

**Comment by AndyAyersMS** [comment]
@BruceForstall PTAL
cc @dotnet/jit-contrib 

@dotnet-bot test Windows_NT x86 Checked r2r_jitminopts
@dotnet-bot test Windows_NT x64 Checked r2r_jitminopts

File: src/jit/lclvars.cpp [hunk 1]
```diff
@@ -4100,7 +4100,16 @@ void Compiler::lvaComputeRefCounts(bool isRecompute, bool setSlotNumbers)
             // to do something else to ensure late temps are considered.
             for (lclNum = 0, varDsc = lvaTable; lclNum < lvaCount; lclNum++, varDsc++)
             {
-                assert(varDsc->lvImplicitlyReferenced);
+                const bool isSpecialVarargsParam = varDsc->lvIsParam && raIsVarargsStackArg(lclNum);
```
Comment (jashook) [inline src/jit/lclvars.cpp]: raIsVarargsStackArg is under an x86 ifdef, I am worried this special case may not be correctly marked on other platforms.

Comment (jashook) [reply src/jit/lclvars.cpp]: For example on x64 if there are 5 args the last arg would not be correctly marked isSpecialVarargsParam correct?

**Comment by AndyAyersMS** [comment]
If you think there are residual problems please open an issue and/or write up the test case that exposes them. I can imagine that x86 could be a special case because of the differences in calling conventions, but we should resolve this for certain.

**Comment by AndyAyersMS** [comment]
Since there is nothing arm-specific here I'm going to merge this since I want to fix up #19345.

File: src/jit/lclvars.cpp [hunk 2]
```diff
@@ -4118,8 +4127,17 @@ void Compiler::lvaComputeRefCounts(bool isRecompute, bool setSlotNumbers)
...
-            varDsc->lvTracked              = 0;
+
+            // Special case for some varargs params ... these must
+            // remain unreferenced.
+            const bool isSpecialVarargsParam = varDsc->lvIsParam && raIsVarargsStackArg(lclNum);
+
+            if (!isSpecialVarargsParam)
+            {
+                varDsc->lvImplicitlyReferenced = 1;
+            }
```
Comment (BruceForstall) [inline src/jit/lclvars.cpp]: Is this where it is initialized? Does it need to be:
`varDsc->lvImplicitlyReferenced = isSpecialVarargsParam : 0 : 1;` ?

Comment (AndyAyersMS) [reply src/jit/lclvars.cpp]: The entire local var table is zero initialized when the table is allocated, so we don't need to set `lvImplicitlyReferenced` to zero. For instance, in `lvaInitTypeRef()`

```
    lvaTable         = getAllocator(CMK_LvaTable).allocate<LclVarDsc>(lvaTableCnt);
    size_t tableSize = lvaTableCnt * sizeof(*lvaTable);
    memset(lvaTable, 0, tableSize);
```

**Pull request merged**